src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs

```diff
@@ -30,12 +31,14 @@ internal interface ITestAccessor
         MembershipVersion ITestAccessor.LastMembershipVersion { get; set; }
 
         public CachedGrainLocator(
+            IServiceProvider serviceProvider,
             GrainDirectoryResolver grainDirectoryResolver,
-            IClusterMembershipService clusterMembershipService)
+            IClusterMembershipService clusterMembershipService,
+            IOptions<GrainDirectoryOptions> grainDirectoryOptions)
         {
             this.grainDirectoryResolver = grainDirectoryResolver;
             this.clusterMembershipService = clusterMembershipService;
-            this.cache = new LRUBasedGrainDirectoryCache(GrainDirectoryOptions.DEFAULT_CACHE_SIZE, GrainDirectoryOptions.DEFAULT_MAXIMUM_CACHE_TTL);
+            this.cache = GrainDirectoryCacheFactory.CreateGrainDirectoryCache(serviceProvider, grainDirectoryOptions.Value);
```

Comment:
The issue is that this method can return an `AdaptiveGrainDirectoryCache`, that only make sense when the builtin `LocalGrainDirectory` is used. It will do weird thing if used here.

But I agree that it would be nicer to be able to inject/customize the cache used here.

Reply:
I see, how about add a new function in `GrainDirectoryCacheFactory` to get  `IGrainDirectoryCache` for DI if there is one or return `LRUBasedGrainDirectoryCache` as a default?
this way we keep the current behavior as is but allow injecting custom IGrainDirectoryCache.
something like:  

       public static IGrainDirectoryCache CreateGrainDirectoryCache(IServiceProvider services, GrainDirectoryOptions options)
       {
           var grainDirectoryCache = services.GetService<IGrainDirectoryCache>();
           if (grainDirectoryCache != null)
           {
               return grainDirectoryCache;
           }
           else
           {
               return new LRUBasedGrainDirectoryCache(options.CacheSize, options.MaximumCacheTTL);
           }
       }

Reply:
@benjaminpetit any feedback about the above ?

Reply:
@benjaminpetit  any feedback about the above?

Reply:
@benjaminpetit no worries, welcome back 🙂
I have pushed the new GrainDirectoryCacheFactory function, 'Naming things` sticks again but I did my best to come up with a name and added a summary to explain what it should do, please tell me if the name is not clear or if it can be improved 😉